### DISCUSSION TITLE
Why edger.r script do not use calcNormFactors function

### DISCUSSION THREAD

**Discussion by ericjuo** [biostars/biostar-handbook-issues#213] [Q&A]
Hi everyone,

I am new to RNA-Seq analysis. I've looked into the edger.r script and cross checked with the edger manual. I found that the edgeR manual uses calcNormFactors function to normalize libraries sizes prior to estimating the dispersion, but the edger.r script didn't include this step. Why is that? 

```
> x <- read.delim("TableOfCounts.txt",row.names="Symbol")
> group <- factor(c(1,1,2,2))
> y <- DGEList(counts=x,group=group)
> keep <- filterByExpr(y)
> y <- y[keep,,keep.lib.sizes=FALSE]
> y <- calcNormFactors(y)
> design <- model.matrix(~group)
> y <- estimateDisp(y,design)
```

**Comment by ialbert** [comment]
the script, as implemented uses the  `exactTest` to later detect the differences in the means between two groups 

https://rdrr.io/bioc/edgeR/man/exactTest.html

the `exactTest` performs a library size normalization internally as described on the page above. This is why there is a normalized matrix that it also saves.

But this is a good point, in that if one is using a different method other functions may need to be called.

**Reply by ericjuo** [reply]
Hi @ialbert,

Thank you for the reply. I still have some questions.

In edger.r script, scale for normalizing counts is calculated by
```
# Get the scale of the data
scale = dge$samples$lib.size * dge$samples$norm.factors
```
However, I looked into the output of dge$samples after exactTest function is called, the values in norm.factors column are all 1. It seems library sizes are not normalized yet.
```
group lib.size norm.factors
cond1 112193 1
cond1 182193 1
cond1 123067 1
cond2 237014 1
cond2 158009 1
cond2 196653 1
```
Besides, despite the document of exactTest says that equalizeLibSizes is called to normalize library sizes before hypothesis test, it doesn't replace normalization factor in the DGEList object (dge variable in this case). The equalizeLibSizes document says that it returns a list with components of peudocounts and peuodo.lib.sizes, and both compenents are mainly for internal use by edgeR, not normally for use by users. 

https://rdrr.io/bioc/edgeR/man/equalizeLibSizes.html

The edgeR manual also states that:
 "The pseudo-counts are computed for a specific purpose, and their computation depends on the experimental design as well as the library sizes, so **users are advised not to interpret the psuedo-counts as general-purpose normalized counts**. They are intended to mainly for internal use in the edgeR pipeline."

Therefore, I think calcNormFactors still needs to be implemented for obtaining normalization factor for each libraries.

**Reply by ialbert** [reply]
The scale factors will not be one, at least the ones that you can extract from the matrix. The matrix is rescaled internally. 

The output of the edger.r script shows you the normalized matrix and not the original counts and you can see there how the numbers change.

I never looked closer to the implementation, and one of the endemic problems with statistical packages is that it is never clearly explained what the tool does and it takes a lot of hunting around to figure out the various cases.

I would myself be curious whether the normalization is needed, I always assumed that for this particular test one would not need to normalize explicitly, but for others, a different approach is required.
